matrix_sdk/src/client.rs

```diff
@@ -1417,17 +1466,19 @@ impl Client {
     ///                 }
     ///             }
     ///         }
+    ///
+    ///         false
     ///     })
     ///     .await;
     /// })
     /// ```
     #[instrument(skip(callback))]
-    pub async fn sync_forever<C>(
+    pub async fn sync_with_callback<C>(
         &self,
         sync_settings: SyncSettings<'_>,
         callback: impl Fn(sync_events::Response) -> C,
     ) where
-        C: Future<Output = ()>,
+        C: Future<Output = bool>,
```

Comment:
Hmm, Gtk uses bools so it may be a convention, but an enum does sound tempting. What names would you expect for the enum variants?

```rust
pub enum {
    Continue,
    Stop,
}
```

Does this sound good?

Reply:
Actually Gtk has a struct wrapping a bool: https://docs.rs/glib/0.10.1/glib/source/struct.Continue.html

I think I'd personally not link the name of the enum to this function in particular, rather doing what Gtk does, or `enum Continue { True, False }` / maybe `enum LoopCtrl { Continue, Break }`.

Reply:
Right, I was talking about the C implementation but looking at what Rust does is indeed more useful. :sweat_smile: 

I picked the second option, unless anyone has more suggestions or objections I'll merge this once CI is satisfied.